proof-of-concept/ast-builder.rkt

```diff
@@ -217,380 +259,394 @@
+                     (? (stx-atom? "="))
+                     (? (stx-many? 'expression) e-stx))
+               (local [(define ids (helper pe-stx))
+                       (define id-list (or (and (i-matrix? ids) (map first (i-matrix-rows ids)))
+                                           (list ids)))] ; can get single assignments too
+                 (i-assn-decl id-list (helper e-stx)))]
+
+              ; postfix_expression
+              [(list (? (stx-atom? 'postfix_expression))
+                     (? (stx-many? 'primary_expression) pe-stx))
+               (helper pe-stx)]
+              [(list (? (stx-atom? 'postfix_expression))
+                     (? (stx-many? 'array_expression) ae-stx))
+               (helper ae-stx)]
+
+              ; array_expression
+              [(list (? (stx-atom? 'array_expression))
+                     (? (stx-many? 'typed_identifier) ti-stx)
+                     (? (stx-atom? "("))
+                     (? (stx-many? 'index_expression_list) iel-stx)
+                     (? (stx-atom? ")")))
+               (i-app (helper ti-stx) (helper iel-stx))]
+
+              ; index_expression_list
+              [(list (? (stx-atom? 'index_expression_list))
+                     (? (stx-many? 'index_expression) ie-stx))
+               (list (helper ie-stx))]
+              [(list (? (stx-atom? 'index_expression_list))
+                     (? (stx-many? 'index_expression_list) iel-stx)
+                     (? (stx-atom? ","))
+                     (? (stx-many? 'index_expression) ie-stx))
+               (append (helper iel-stx) (list (helper ie-stx)))]
+
+              ; index_expression
+              [(list (? (stx-atom? 'index_expression))
+                     (? (stx-atom? ":")))
+               (i-null)]
+              [(list (? (stx-atom? 'index_expression))
+                     (? (stx-many? 'expression) e-stx))
+               (helper e-stx)]
+
+              ; primary_expression
+              [(list (? (stx-atom? 'primary_expression))
+                     (? (stx-many? 'typed_identifier) ti-stx))
+               (helper ti-stx)]
+              [(list (? (stx-atom? 'primary_expression))
+                     (? syntax? literal)) ; string literal
+               (parse-literal literal)]
+              [(list (? (stx-atom? 'primary_expression))
+                     (? (stx-atom? "("))
+                     (? (stx-many? 'expression) e-stx)
+                     (? (stx-atom? ")")))
+               (helper e-stx)]
+              [(list (? (stx-atom? 'primary_expression))
+                     (? (stx-atom? "["))
+                     (? (stx-atom? "]")))
+               (i-matrix empty)]
+              [(list (? (stx-atom? 'primary_expression))
+                     (? (stx-atom? "["))
+                     (? (stx-many? 'array_list) al-stx)
+                     (? (stx-atom? "]")))
+               (local [(define (extract-rows-helper flat-vals rsf acc-res)
```

Comment:
Formatting was a mistake... just see this, array_list, array_element and the extra type case definitions